[BEAM-3217] add HadoopInputFormatIO integration test using DBInputFormat

lgajowy opened the pull request:
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

@chamikaramj could you take a look? What do you think about the idea of using JdbcIO to write the data? There is no write in HadoopInputFormatIO. 

Currently the test is working only for small datasets (100 000 rows, which is about 3,43 MB). For 1 000 000 rows the test is flaky (different hashes get generated). I'm investigating the cause now. After that I'm planning to add the a large dataset hash. 

Also, last thing: large scale scenarios (e.g 40 000 000 rows, which is 10gb according to dataflow's estimation) can take quite a long time to run. Only the write pipline executes for more than 35 minutes. The JdbcIO.write() seems to be the bottleneck, as it is done sequentially - one-row inserts, one after another. I think grouping and then inserting "batches" of rows to the database will speed things up, though I don't know how much. Should I do this or the execution time is bearable? Are there some other optimisations I might want to consider?

PS: Happy new year! 🎊 :)

chamikaramj commented on sdks/java/io/hadoop/input-format/src/test/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIOIT.java:
```diff
@@ -0,0 +1,133 @@
+        .apply("Get values only", Values.<TestRowDBWritable>create())
+        .apply("Values as string", ParDo.of(new SelectNameFn()))
+        .apply("Calculate hashcode", Combine.globally(new HashingFn()))
+        .apply(Reshuffle.<String>viaRandomKey());
```
<!--new_thread; commit:138b0e00d1ea86cd27700083ee657e8c5abae4a6; resolved:0-->
Add a comment explaining why we need this reshuffle and add a link to the JIRA.

chamikaramj replied:
<!--thread_id:cc_159567125_t; commit:138b0e00d1ea86cd27700083ee657e8c5abae4a6; resolved:1-->
<!--section:context-quote-->
> **chamikaramj** wrote:
> Add a comment explaining why we need this reshuffle and add a link to the JIRA.

<!--section:body-->
Done.

chamikaramj commented on sdks/java/io/hadoop/input-format/src/test/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIOIT.java:
```diff
@@ -0,0 +1,133 @@
+    writePipeline.run().waitUntilFinish();
+
+    PCollection<String> consolidatedHashcode = readPipeline
+        .apply("Read using DBInputFormat", HadoopInputFormatIO
```
<!--new_thread; commit:138b0e00d1ea86cd27700083ee657e8c5abae4a6; resolved:0-->
Prob. rename to "Read using HadoopInputFormat" since that explains better what we are testing here.

chamikaramj replied:
<!--thread_id:cc_159567126_t; commit:138b0e00d1ea86cd27700083ee657e8c5abae4a6; resolved:1-->
<!--section:context-quote-->
> **chamikaramj** wrote:
> Prob. rename to "Read using HadoopInputFormat" since that explains better what we are testing here.

<!--section:body-->
Done.

chamikaramj commented on sdks/java/io/hadoop/input-format/src/test/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIOIT.java:
```diff
@@ -0,0 +1,133 @@
+  public void writeThenReadUsingDBInputFormat() {
+    writePipeline.apply("Generate sequence", GenerateSequence.from(0).to(numberOfRows))
+        .apply("Produce db rows", ParDo.of(new DeterministicallyConstructTestRowFn()))
+        .apply(JdbcIO.<TestRow>write()
```
<!--new_thread; commit:138b0e00d1ea86cd27700083ee657e8c5abae4a6; resolved:0-->
Add a label, For example "Write using JDBCIO")

chamikaramj replied:
Writing using JDBCIO is fine. Probably your pipeline is slow since you do all the writing From a single worker. Consider creating a PCollection of seed objects (splits) followed by a Reshuffle() followed by writing.

chamikaramj replied:
<!--thread_id:cc_159567128_t; commit:138b0e00d1ea86cd27700083ee657e8c5abae4a6; resolved:1-->
<!--section:context-quote-->
> **chamikaramj** wrote:
> Writing using JDBCIO is fine. Probably your pipeline is slow since you do all the writing From a single worker. Consider creating a PCollection of seed objects (splits) followed by a Reshuffle() followed by writing.

<!--section:body-->
Done.

chamikaramj commented on sdks/java/io/hadoop/input-format/src/test/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIOIT.java:
```diff
@@ -0,0 +1,133 @@
+  }
+
+  @Test
+  public void writeThenReadUsingDBInputFormat() {
```
<!--new_thread; commit:138b0e00d1ea86cd27700083ee657e8c5abae4a6; resolved:0-->
Rename to readUsingHadoopInputFormat() ? (since write part is done usning JDBCIO).

chamikaramj replied:
<!--thread_id:cc_159567129_t; commit:138b0e00d1ea86cd27700083ee657e8c5abae4a6; resolved:1-->
<!--section:context-quote-->
> **chamikaramj** wrote:
> Rename to readUsingHadoopInputFormat() ? (since write part is done usning JDBCIO).

<!--section:body-->
Done.

chamikaramj commented on sdks/java/io/common/src/test/java/org/apache/beam/sdk/io/common/DatabaseTestHelper.java:
```diff
@@ -0,0 +1,78 @@
+    }
+  }
+
+  public static void cleanUpDataTable(DataSource dataSource, String tableName)
```
<!--new_thread; commit:138b0e00d1ea86cd27700083ee657e8c5abae4a6; resolved:0-->
deleteTable ?

chamikaramj replied:
<!--thread_id:cc_159567130_t; commit:138b0e00d1ea86cd27700083ee657e8c5abae4a6; resolved:1-->
<!--section:context-quote-->
> **chamikaramj** wrote:
> deleteTable ?

<!--section:body-->
Done.

chamikaramj commented on sdks/java/io/common/src/test/java/org/apache/beam/sdk/io/common/DatabaseTestHelper.java:
```diff
@@ -0,0 +1,78 @@
+    return dataSource;
+  }
+
+  public static void createDataTable(DataSource dataSource, String tableName)
```
<!--new_thread; commit:138b0e00d1ea86cd27700083ee657e8c5abae4a6; resolved:0-->
createTable ?

chamikaramj replied:
<!--thread_id:cc_159567131_t; commit:138b0e00d1ea86cd27700083ee657e8c5abae4a6; resolved:1-->
<!--section:context-quote-->
> **chamikaramj** wrote:
> createTable ?

<!--section:body-->
Done.

chamikaramj commented:
cc: @jbonofre @iemejia

lgajowy commented:
@chamikaramj thank you for your suggestions.  @jbonofre @iemejia could you also take a look? I also added io-it-suite-local profile that was missing and jenkins job definition.

>Writing using JDBCIO is fine. Probably your pipeline is slow since you do all the writing From a single worker. Consider creating a PCollection of seed objects (splits) followed by a Reshuffle() followed by writing.

I added only the reshuffle and it seems to be a little bit helpful. I didn't optimise it further due to a  problem: different "consolidatedHashes" get calculated for each test run for datasets bigger than 600 000 rows. This makes it unable to determine hash for a large scale dataset (eg. 40 000 000 rows). The amount of read and written rows is the same. Also, as I checked, the database content seems to be all right. I can create a JIRA for that after you review this PR and agree that this behavior is odd, ok?

600 000 is approx. 160 MB. I wouldn't call that a large scale test but I think it is something we can start with and then increase the scale and optimize it gradually if needed and if possible (e.g. after tackling the hash calculaction problem i described). What do you think?

lgajowy commented on sdks/java/io/hadoop/input-format/pom.xml:
```diff
@@ -27,6 +27,197 @@
   <name>Apache Beam :: SDKs :: Java :: IO :: Hadoop :: input-format</name>
   <description>IO to read data from data sources which implement Hadoop Input Format.</description>
 
+  <profiles>
```
For now I simply duplicated the code from jdbc io, applying necessary changes. I think it can be improved later, see: https://issues.apache.org/jira/browse/BEAM-3440

chamikaramj commented:
Might be worth filing a JIRA for the JDBC issue in case it's a bug in the sink.

I'm fine with enabling this tests for a smaller Dataset and increasing the size later after fixes.

chamikaramj commented:
Run seed job

chamikaramj commented:
Run Java HadoopInputFormatIO Performance Test

chamikaramj commented:
Any idea why following failed ?

https://builds.apache.org/job/beam_PerformanceTests_HadoopInputFormatIO_IT/1/console

lgajowy commented:
Run Java HadoopInputFormatIO Performance Test

lgajowy commented:
Run Java HadoopInputFormatIO Performance Test

lgajowy commented:
Run seed job

lgajowy commented:
Run Java HadoopInputFormatIO Performance Test

lgajowy commented:
Run seed job

lgajowy commented:
Run Java HadoopInputFormatIO Performance Test

lgajowy commented:
Run Java HadoopInputFormatIO Performance Test

lgajowy commented:
Run Java HadoopInputFormatIO Performance Test

lgajowy commented:
Run seed job

lgajowy commented:
Run Java HadoopInputFormatIO Performance Test

lgajowy commented:
Run Java HadoopInputFormatIO Performance Test

lgajowy commented:
Run Java HadoopInputFormatIO Performance Test

lgajowy commented:
Run Java HadoopInputFormatIO Performance Test

lgajowy commented:
>Any idea why following failed ?
> https://builds.apache.org/job/beam_PerformanceTests_HadoopInputFormatIO_IT/1/console

Judging from the logs you attached the kubeconfig location is wrong. In the new commit i tried to overwrite it with value I got from our local jenkins setup hoping that it's some default value. No luck. 

I think someone with access to jenkins is needed - we need to know the path to kubeconfig to set it up correctly. @jbonofre can you help?

lgajowy commented:
Run seed job

lgajowy commented:
Run Java HadoopInputFormatIO Performance Test

lgajowy commented:
Run Java HadoopInputFormatIO Performance Test

lgajowy commented:
Run Java HadoopInputFormatIO Performance Test

lgajowy commented:
Run Java HadoopInputFormatIO Performance Test

lgajowy commented:
Run Java HadoopInputFormatIO Performance Test

lgajowy commented:
Run Java HadoopInputFormatIO Performance Test

lgajowy commented:
Run seed job

lgajowy commented:
Run Java HadoopInputFormatIO Performance Test

lgajowy commented:
Run Java HadoopInputFormatIO Performance Test

lgajowy commented:
I also tried to set different path to kubernetes scripts - analogous to the one that was in JDBC tests: https://builds.apache.org/view/A-D/view/Beam/job/beam_PerformanceTests_JDBC/215/console

also no luck, because i got some "permission denied error" even earlier (the path didn't even matter that early): https://builds.apache.org/job/beam_PerformanceTests_HadoopInputFormatIO_IT/16/console

@jbonofre can you help in diagnosing what is going on?

lgajowy commented:
Run seed job

lgajowy commented:
Run Java HadoopInputFormatIO Performance Test

lgajowy commented:
Run Java HadoopInputFormatIO Performance Test

lgajowy commented:
Run Java HadoopInputFormatIO Performance Test

lgajowy commented:
Removed jenkins job due to the reasons described in [4392](https://github.com/apache/beam/pull/4392#issuecomment-358304303) pull request. The job should be added in a separate PR after the problems are solved.

@chamikaramj could you take a look?

chamikaramj commented:
Can you merge and resolve conflicts ?

LGTM other than that.

chamikaramj commented:
LGTM. Thanks.

lukecwik commented:
This broke the gradle build:
/home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_GradleBuild/src/sdks/java/io/hadoop/input-format/src/test/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIOIT.java:20: error: package org.apache.beam.sdk.io.common does not exist
import static org.apache.beam.sdk.io.common.TestRow.DeterministicallyConstructTestRowFn;
                                           ^
/home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_GradleBuild/src/sdks/java/io/hadoop/input-format/src/test/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIOIT.java:20: error: static import only from classes and interfaces
import static org.apache.beam.sdk.io.common.TestRow.DeterministicallyConstructTestRowFn;
^

Filed: https://issues.apache.org/jira/browse/BEAM-3496
